File: backend/src/main/java/cn/etarch/mao/harness/core/ToolAttachmentLoader.java
package cn.etarch.mao.harness.core;

import cn.etarch.mao.harness.tool.ImageFileSupport;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.extern.slf4j.Slf4j;

import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Loads tool image attachments from persisted message metadata.
 */
@Slf4j
public final class ToolAttachmentLoader {

    private ToolAttachmentLoader() {
    }

    public static void loadFromMetadata(String toolCallId, String metadataJson,
                                        Map<String, ToolAttachment> target, ObjectMapper objectMapper) {
        if (toolCallId == null || metadataJson == null || metadataJson.isBlank() || target == null) {
            return;
        }
        try {
            JsonNode root = objectMapper.readTree(metadataJson);
            JsonNode attachments = root.get("attachments");
            if (attachments == null || !attachments.isArray() || attachments.isEmpty()) {
                return;
            }
            JsonNode first = attachments.get(0);
            String mime = first.path("mime").asText(null);
            String path = first.path("path").asText(null);
            String dataUri = first.path("data_uri").asText(null);
            if (!ImageFileSupport.isImageMime(mime) || dataUri == null || dataUri.isBlank()) {
                return;
            }
            target.put(toolCallId, ToolAttachment.builder()
                    .mime(mime)
                    .path(path)
                    .dataUri(dataUri)
                    .build());
        } catch (Exception e) {
            log.warn("Failed to parse tool attachment metadata for toolCallId={}", toolCallId, e);
        }
    }

    public static Map<String, ToolAttachment> loadAllFromMessages(
            Iterable<cn.etarch.mao.session.entity.Message> messages, ObjectMapper objectMapper) {
        Map<String, ToolAttachment> attachments = new LinkedHashMap<>();
        for (cn.etarch.mao.session.entity.Message msg : messages) {
            if (!"TOOL".equals(msg.getRole()) || msg.getToolCallId() == null) {
                continue;
            }
            loadFromMetadata(msg.getToolCallId(), msg.getMetadata(), attachments, objectMapper);
        }
        return attachments;
    }
}

File: backend-ts/src/harness/core/tool-attachment-loader.ts
import type { Message } from '../deps.js';
import type { ToolAttachment } from './tool-attachment.js';
import { ImageFileSupport } from '../tool/image-file-support.js';
import { harnessLog } from '../log.js';

export const ToolAttachmentLoader = {
  loadFromMetadata(
    toolCallId: string | null | undefined,
    metadataJson: string | null | undefined,
    target: Map<string, ToolAttachment> | Record<string, ToolAttachment>,
  ): void {
    if (!toolCallId || !metadataJson || metadataJson.trim() === '') return;
    try {
      const root = JSON.parse(metadataJson) as Record<string, unknown>;
      const attachments = root.attachments;
      if (!Array.isArray(attachments) || attachments.length === 0) return;
      const first = attachments[0] as Record<string, unknown>;
      const mime = first.mime as string | undefined;
      const filePath = first.path as string | undefined;
      const dataUri = first.data_uri as string | undefined;
      if (!ImageFileSupport.isImageMime(mime) || !dataUri) return;
      const att: ToolAttachment = { mime, path: filePath, dataUri };
      if (target instanceof Map) target.set(toolCallId, att);
      else (target as Record<string, ToolAttachment>)[toolCallId] = att;
    } catch (e) {
      harnessLog('warn', `Failed to parse tool attachment metadata for toolCallId=${toolCallId}`, e);
    }
  },

  loadAllFromMessages(messages: Iterable<Message>): Record<string, ToolAttachment> {
    const attachments: Record<string, ToolAttachment> = {};
    for (const msg of messages) {
      if (msg.role !== 'TOOL' || !msg.toolCallId) continue;
      this.loadFromMetadata(msg.toolCallId, msg.metadata, attachments);
    }
    return attachments;
  },
};
